bfvmm/include/hve/arch/intel_x64/vcpu.h

```diff
@@ -347,61 +351,36 @@ class vcpu : public bfvmm::vcpu
     VIRTUAL void execute_wrcr4();
 
     //--------------------------------------------------------------------------
-    // CPUID
+    // vCPU CPUID Interface Implementation
     //--------------------------------------------------------------------------
 
-    /// Add CPUID Handler
-    ///
-    /// Add a delegate to handle a CPUID VM exit. Your handler should always
-    /// return false unless you need to override the default behavior of the
-    /// base hypervisor. For more information, please see the CPUID VM exit
-    /// handler for details.
-    ///
-    /// @expects
-    /// @ensures
-    ///
-    /// @param leaf the leaf to call d on
-    /// @param d the delegate to call on the vm exit
-    ///
-    VIRTUAL void add_cpuid_handler(
-        cpuid_handler::leaf_t leaf, const handler_delegate_t &d);
+    /// @cond
 
-    /// Add Emulate
-    ///
-    /// Emulate the exeuction of the CPUID instruction. For more information
-    /// please see the CPUID VM exit handler for details. Generally speaking,
-    /// you should typically use add_handler() instead.
-    ///
-    /// @expects
-    /// @ensures
-    ///
-    /// @param leaf the leaf to call d on
-    /// @param d the delegate to call on the vm exit
-    ///
-    VIRTUAL void add_cpuid_emulator(
-        cpuid_handler::leaf_t leaf, const handler_delegate_t &d);
+    void cpuid_add_handler(
+        vcpu_cpuid_interface::leaf_t leaf, const handler_delegate_t &d) OVERRIDE
+    { m_cpuid_handler.add_handler(leaf, d); }
```

Comment:
these shouldn't be here now that we are moving everything to the interface. The implementation should look like this when we are done.

```
class vcpu : 
    public bfvmm::vcpu,
    public bfvmm::interface::cpuid,
    ....
{ 
public:
    vcpu::vcpu(id) : 
        bfvmm::vcpu(id),
        bfvmm::interface::cpuid(this),
        ....
    { }
};
````

And then each interface actually implements the stable API (using both headers and source)

````
class cpuid
{
public:
    cpuid(vcpu_t *vcpu);

    // stable APIs go here

private:

    // Implementation specific details go here
};
````

Pure virtual base class functions are only needed when you intend to enforce multiple inheritance. In general, however, C++ Concepts are a more prefered mechanism that accomplishes the same thing. In this specific case, the `cpuid` APIs are Intel specific, so there is no reason to provide pure virtual functions as nothing else will inherit this. 

At some point when we attempt to add support for AMD, we will likely end up with some code that is shared in which case it will need to be moved to a non-Intel location, but the APIs and how they are used will still remain stable so long as they adhere to the concept the interface defines. 

Here is an example of a C++ concept (there are many):
http://www.enseignement.polytechnique.fr/informatique/INF478/docs/Cpp/en/cpp/concept/Allocator.html

This interface can be implemented by anything and with whatever private implementation it wants, and the code works without modification so long as the concept is adhered to. Another way to state this is, to create a stable interface, all we have to do is define what the interface is and its requirements. Any archiectural implementation from that point on simply needs to implement this interface to work, and from a stability point of view, modifications to the concept are what must be guarded, knowing that any changes to the concept will effect everyone and so they should only be done if absolutely needed. 

The only issue with concepts today in C++ is that if the code doesn't adhere to the concept, the only way you will know is when you attempt to compile and something doesn't add up which results in weird compiler errors. This is however being address in C++20, and we shouldn't attempt to reinvent this:
https://en.cppreference.com/w/cpp/language/constraints

This is further being expanded in C++20 with the use of contracts:
https://en.cppreference.com/w/cpp/language/attributes/contract

All of this can be used to further define and enforce the stable API once the features are available, and this is the direction the language is taking. Until then, our goal should be to do the same thing the Standard Library does which is to define the concept for each interface and then stick to it. Modifications should only be made if they are absolutely needed to ensure downstream users are not effected. 

One final note about pure virtual functions. The main issue with this approach is it significantly increase the size of the vTable for the object. One main reason that concepts are used over pure virtual interfaces is these interfaces were only meant for functions that are needed by a base implemention. For example, if you have a class that defines an Animal and you then want to create a Dog and a Cat and a function for Talk (the anmial makes a noise), you do not need to inherit and override a base function in the Animal class unless you intend to store only Animals and call the Talk function from your list of Animals. If, however, you intend to use the Cat and Dog classes as they are, you can call the Talk function from any other class through the use of concepts. In fact, there is no need to even define the Animal class in the first place. To take this a step futher, even if you plan to store only Animals and call the Talk function from your base Animal, you still do not need to inherit and override as Type Erasure can be used to implement concepts as well. 

This doesn't mean that inheritance is bad or that it doesn't have its place, it simply means that if concepts can be used instead of inheritance, they should as they prevent the class from ballooning out of propotion in size, and they also prevent the need for a vTable entry which ultanitemly prevents inlining and adds an additional function call when it is not needed to accomplish the same thing. 

P.S. even if a design like this way used, you really should be using "final" instead of "override" unless you intend the APIs to be overriden by the APIs users.